keeper/components/dispatcher_stats.go

```diff
@@ -0,0 +1,183 @@
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN "AS IS" BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package components
+
+import (
+	"context"
+	"fmt"
+
+	"cosmossdk.io/math"
+
+	"orbiter.dev/types"
+)
+
+// updateStats updates all the statistics the module keep track of.
+// CONTRACT: transferAttr and orbit are not nil pointers.
+func (d *DispatcherComponent) updateStats(
+	ctx context.Context,
+	transferAttr *types.TransferAttributes,
+	orbit *types.Orbit,
+) error {
+	attr, err := orbit.CachedAttributes()
+	if err != nil {
+		return err
+	}
+
+	var sourceAttr *types.OrbitID
+	if sourceAttr, err = types.NewOrbitID(
+		transferAttr.SourceProtocolID(),
+		transferAttr.SourceCounterpartyID(),
+	); err != nil {
+		return err
+	}
+
+	var destinationAttr *types.OrbitID
+	if destinationAttr, err = types.NewOrbitID(
+		orbit.ProtocolID(),
+		attr.CounterpartyID(),
+	); err != nil {
+		return err
+	}
+
+	// Since incoming denom can be different than the outgoing one,
+	// we have to check here how many amount dispatched types to store.
+	// If the denom is not changed, we can set a single type with incoming
+	// and outgoing amount, which could be different too, but are not part
+	// of the key. If the denom changed, we have to set two values with
+	// a different denom.
+	denomDispatchedAmounts := d.buildDenomDispatchedAmounts(transferAttr)
+
+	for _, dda := range denomDispatchedAmounts {
+		if err := d.updateDispatchedAmountStats(ctx, sourceAttr, destinationAttr, dda.Denom, dda.AmountDispatched); err != nil {
+			return fmt.Errorf("update incoming stats failure: %w", err)
+		}
+	}
+
+	if err := d.updateDispatchedCountsStats(ctx, sourceAttr, destinationAttr); err != nil {
+		return fmt.Errorf("update dispatch counts stats failure: %w", err)
+	}
+
+	return nil
+}
+
+// updateDispatchedAmountStats updates the amount dispatched
+// values on the store. A boolean flag is used to indicate
+// if the amount to be added is an incoming or outgoing amount.
+// It is important to keep track of incoming and outgoing
+// information because fees, swaps, or other actions can change
+// the coins delivered to the destination chain.
+func (d *DispatcherComponent) updateDispatchedAmountStats(
+	ctx context.Context,
+	sourceInfo *types.OrbitID,
+	destinationInfo *types.OrbitID,
+	denom string,
+	newAmountDispatched types.AmountDispatched,
+) error {
+	amountDispatched := d.GetDispatchedAmount(
+		ctx,
+		*sourceInfo,
+		*destinationInfo,
+		denom,
+	)
+
+	if newAmountDispatched.Incoming.IsPositive() {
+		amountDispatched.Incoming = amountDispatched.Incoming.Add(newAmountDispatched.Incoming)
+	}
+	if newAmountDispatched.Outgoing.IsPositive() {
+		amountDispatched.Outgoing = amountDispatched.Outgoing.Add(newAmountDispatched.Outgoing)
+	}
```

Comment:
should this rather check for the opposite case and return an error if some value is negative? that would be break a basic assumption about the model and would deserve to be marked as an error no?